pkg/newlog/cmd/newlogd.go

```diff
@@ -0,0 +1,1281 @@
+				nmsg = msg2[1]
+			}
+		}
+		if nsev != "" || nmsg != "" {
+			loginfo.Level = nsev
+			loginfo.Msg = nmsg
+			ok = true
+		}
+	} else {
+		msgStr := myStr1[0]
+		if string(msgStr[len(msgStr)-1]) != "}" {
+			msgStr += "}"
+		}
+		loginfo, ok = agentlog.ParseLoginfo(msgStr)
+		if ok {
+			loginfo.Msg = myStr1[1]
+		}
+	}
+	if !ok {
+		loginfo.Level = "info"
+		loginfo.Msg = content
+	}
+	return loginfo
+}
+
+func getSourceFromMsg(msg string) (string, string, string) {
+	var source, content string
+	lines := strings.SplitN(msg, ";", 2)
+	if len(lines) != 2 {
+		return source, "", content
+	}
+	content = lines[1]
+	lines2 := strings.Split(lines[0], ",")
+	n := len(lines2)
+	if n < 2 {
+		return source, "", content
+	}
+	return lines2[n-1], lines2[n-2], content
+}
+
+func startTmpfile(dirname, filename string) *os.File {
+	tmpFile, err := ioutil.TempFile(dirname, filename)
+	if err != nil {
+		log.Fatal(err)
+	}
+	err = tmpFile.Chmod(0600)
+	if err != nil {
+		log.Fatal(err)
+	}
+	return tmpFile
+}
+
+func remNonPrintable(str string) string {
+	var re = regexp.MustCompile(ansi)
+	myStr := re.ReplaceAllString(str, "")
+	myStr = strings.Trim(myStr, "\r")
+	return strings.Trim(myStr, "\n")
+}
+
+// writelogFile - a goroutine to format and write log entries into dev/app logfiles
+func writelogFile(logChan <-chan inputEntry, moveChan chan fileChanInfo) {
+
+	if _, err := os.Stat(collectDir); os.IsNotExist(err) {
+		if err := os.MkdirAll(collectDir, 0755); err != nil {
+			log.Fatal(err)
+		}
+	}
+
+	// move and gzip the existing logfiles first
+	logmetaData = findMovePrevLogFiles(moveChan)
+
+	// new logfile
+	devlogFile := startTmpfile(collectDir, devPrefix)
+	defer devlogFile.Close()
+
+	var fileinfo fileChanInfo
+	var devStats statsLogFile
+
+	devSourceBytes = make(map[string]uint64)
+	appStatsMap = make(map[string]statsLogFile)
+	checklogTimer := time.NewTimer(logfileDelay * time.Second)
+	devStats.file = devlogFile
+
+	// write the first log metadata to the first line of the logfile, will be extracted when
+	// doing gzip conversion. further log file's metadata is handled inside 'trigMoveToGzip()'
+	_, err := devStats.file.WriteString(logmetaData + "\n")
+	if err != nil {
+		log.Fatal(err)
+	}
+
+	timeIdx := 0
+	for {
+		select {
+		case <-checklogTimer.C:
+			timeIdx++
+			checkLogTimeExpire(fileinfo, &devStats, moveChan)
+			checklogTimer = time.NewTimer(15 * time.Second) // check the file time limit every 15 seconds
+			if timeIdx%120 == 0 {                           // every half an hour check space left
+				spaceAvailMB = getAvailableSpace()
+			}
+
+		case entry := <-logChan:
+			appuuid := checkAppEntry(&entry)
+			timeS := getPtypeTimestamp(entry.timestamp)
+			mapLog := logs.LogEntry{
+				Severity:  entry.severity,
+				Source:    entry.source,
+				Content:   entry.content,
+				Iid:       entry.pid,
+				Filename:  entry.filename,
+				Msgid:     msgIDCounter,
+				Function:  entry.function,
+				Timestamp: &timeS,
```

Comment:
Do you need this timestamp to be in protobuf format as opposed to a string with RFC3339? What does the receiver in the controller do with the protobuf-encoded timestamp?

Reply:
And why use a struct generated from protoc? Can't you define a normal go struct for this? I don't know what other public fields (apart from the lock) which you will json encode from the protoc generated struct.

Reply:
on the controller side, basically this: time.Unix(entry.Timestamp.Seconds, int64(entry.Timestamp.Nanos)) after the json Unmarshal. again, this is to reuse all the existing APIs of logging for EVE/controller with no change.